### ISSUE TITLE
Diacritic folding should be language-aware

### ISSUE THREAD

**Issue by 1ec5** [mapbox/carmen#610]
When the language of a feature is known, diacritic-folding rules specific to that language should be applied.

### Context

For Latin alphabetic text, remove-diacritics.js performs a basic form of diacritic folding in which diacritics are removed. However, many languages establish more sophisticated transformations between the conventional orthography and the basic English alphabet. For example, in German, “über” is equivalent to “ueber”, not “uber”. Yet “über” and “uber” return identical results while “ueber” returns very different results.

### Next Steps

This is a task that [ICU](http://userguide.icu-project.org/transforms/general)’s `Latin-ASCII` transliterator excels at. It may be worth investigating the mappings used by that transliterator.

/cc @apendleton @boblannon

**Comment by apendleton** [comment]
I don't think this is really doable given our current architecture. We could do language-specific translations on text on the indexer side, but we don't necessarily know on the query side what language query text is in (the `language` flag is optional), and it's critical for successful lookup that the same transformations be applied to query text as for input text so that strings can be matched successfully. If a user queries for "Köln" without a language flag and we apply a language-agnostic transformation turning into "Koln," but our German text has been indexed as "Koeln" because we applied a German-specific transformation, lookup will fail.

The German case in particular is already on my radar, and I think it's better handled entirely at indexing time, by indexing synonyms that have the diacritical marks substituted (so, indexing both "Köln"=>"Koln" and "Koeln" as synonymous German strings for the same feature) -- this is a change that need only be made on the indexing side, where the language is much more likely to be known, and also keeps from requiring the introduction of text transformations whose performance characteristics are outside our control into the query-time critical path where latency is an especially important consideration.

If there are other languages that have similar atypical diacritical-mark-handling rules, I'd prefer to also handle those case-by-case. If it turns out this is a broad, systematic problem I'd be open to introducing a heavy-weight dependency like ICU, but until that's apparent I'd just as well not, and even then I'd want to confine it to indexing and leave the query-time transformation as light-weight as possible.

**Comment by mtmail** [comment]
Adding another German test case since there is no town called "Koln": There's the town of Münster (plus dozens smaller places https://de.wikipedia.org/wiki/M%C3%BCnster), which is different from the various Munster (https://de.wikipedia.org/wiki/Munster_(%C3%96rtze) though more outside Germany https://en.wikipedia.org/wiki/Munster_(disambiguation))

Your approach looks how to handle that during index and run time looks good.
